ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/electra/util/AttestationUtilElectra.java

```diff
@@ -93,4 +115,162 @@ public AttestationData getGenericAttestationData(
+    }
+
+    // single attestation flow
+
+    // 1. verify signature first
+    // 2. verify call getSingleAttestationAggregationBits which also validates the validatorIndex
+    // and the committee against the state
+    // 3. convert attestation inside ValidatableAttestation to AttestationElectra
+    // 4. set the indexed attestation into ValidatableAttestation
+    // 5. set the attestation as valid indexed attestation
+
+    return validateSingleAttestationSignature(
+            fork,
+            state,
+            attestation.getAttestation().toSingleAttestationRequired(),
+            blsSignatureVerifier)
+        .thenApply(
+            result -> {
+              if (result.isSuccessful()) {
+                final SingleAttestation singleAttestation =
+                    attestation.getAttestation().toSingleAttestationRequired();
+                final IndexedAttestation indexedAttestation =
+                    getIndexedAttestationFromSingleAttestation(singleAttestation);
+
+                final SszBitlist singleAttestationAggregationBits =
+                    getSingleAttestationAggregationBits(state, singleAttestation);
+
+                final Attestation convertedAttestation =
+                    convertSingleAttestationToAggregated(
+                        singleAttestation, singleAttestationAggregationBits);
+
+                attestation.convertToAggregatedFormatFromSingleAttestation(convertedAttestation);
+                attestation.saveCommitteeShufflingSeedAndCommitteesSize(state);
+                attestation.setIndexedAttestation(indexedAttestation);
+                attestation.setValidIndexedAttestation();
+              }
+              return result;
+            })
+        .exceptionallyCompose(
+            err -> {
+              if (err.getCause() instanceof IllegalArgumentException) {
+                LOG.debug("on_attestation: Attestation is not valid: ", err);
+                return SafeFuture.completedFuture(
+                    AttestationProcessingResult.invalid(err.getMessage()));
+              } else {
+                return SafeFuture.failedFuture(err);
+              }
+            });
+  }
+
+  private Attestation convertSingleAttestationToAggregated(
+      final SingleAttestation singleAttestation,
+      final SszBitlist singleAttestationAggregationBits) {
+    final AttestationElectraSchema attestationElectraSchema =
+        schemaDefinitions.getAttestationSchema().toVersionElectra().orElseThrow();
+
+    return attestationElectraSchema.create(
+        singleAttestationAggregationBits,
+        singleAttestation.getData(),
+        singleAttestation.getAggregateSignature(),
+        attestationElectraSchema
+            .getCommitteeBitsSchema()
+            .orElseThrow()
+            .ofBits(singleAttestation.getFirstCommitteeIndex().intValue()));
+  }
+
+  private SafeFuture<AttestationProcessingResult> validateSingleAttestationSignature(
```

Comment:
I guess we could refactor `AttestationUtil::isValidIndexedAttestationAsync` and reuse the signature verification